src/librustc_mir/dataflow/impls/borrows.rs

```diff
@@ -0,0 +1,179 @@
+                                          location_map: FxHashMap(),
+                                          region_map: FxHashMap(), };
+        visitor.visit_mir(mir);
+        return Borrows { tcx: tcx,
+                         mir: mir,
+                         borrows: visitor.idx_vec,
+                         location_map: visitor.location_map,
+                         region_map: visitor.region_map, };
+
+        struct GatherBorrows<'tcx> {
+            idx_vec: IndexVec<BorrowIndex, BorrowData<'tcx>>,
+            location_map: FxHashMap<Location, BorrowIndex>,
+            region_map: FxHashMap<Region<'tcx>, FxHashSet<BorrowIndex>>,
+        }
+        impl<'tcx> Visitor<'tcx> for GatherBorrows<'tcx> {
+            fn visit_rvalue(&mut self,
+                            rvalue: &mir::Rvalue<'tcx>,
+                            location: mir::Location) {
+                if let mir::Rvalue::Ref(region, kind, ref lvalue) = *rvalue {
+                    let borrow = BorrowData {
+                        location: location, kind: kind, region: region, lvalue: lvalue.clone(),
+                    };
+                    let idx = self.idx_vec.push(borrow);
+                    self.location_map.insert(location, idx);
+                    let mut borrows = self.region_map.entry(region).or_insert(FxHashSet());
+                    borrows.insert(idx);
+                }
+            }
+        }
+    }
+
+    pub fn borrows(&self) -> &IndexVec<BorrowIndex, BorrowData<'tcx>> { &self.borrows }
+
+    pub fn location(&self, idx: BorrowIndex) -> &Location {
+        &self.borrows[idx].location
+    }
+}
+
+impl<'a, 'tcx> BitDenotation for Borrows<'a, 'tcx> {
+    type Idx = BorrowIndex;
+    fn name() -> &'static str { "borrows" }
+    fn bits_per_block(&self) -> usize {
+        self.borrows.len()
+    }
+    fn start_block_effect(&self, _sets: &mut BlockSets<BorrowIndex>)  {
+        // no borrows of code extents have been taken prior to
+        // function execution, so this method has no effect on
+        // `_sets`.
+    }
+    fn statement_effect(&self,
+                        sets: &mut BlockSets<BorrowIndex>,
+                        location: Location) {
+        let block = &self.mir.basic_blocks().get(location.block).unwrap_or_else(|| {
+            panic!("could not find block at location {:?}", location);
+        });
+        let stmt = block.statements.get(location.statement_index).unwrap_or_else(|| {
+            panic!("could not find statement at location {:?}");
+        });
+        match stmt.kind {
+            mir::StatementKind::EndRegion(extent) => {
+                let borrow_indexes = self.region_map.get(&ReScope(extent)).unwrap_or_else(|| {
+                    panic!("could not find BorrowIndexs for code-extent {:?}", extent);
+                });
+
+                for idx in borrow_indexes { sets.kill(&idx); }
+            }
+
+            mir::StatementKind::Assign(_, ref rhs) => {
```

Comment:
The thing is, when we have something like

```Rust
let mut t1 = 1;
let mut t2 = 2;

let mut x = &'α mut t1;
let y = &'β mut *x;

x = &mut y2;
// `*x` is no longer borrowed here (because of the assignment)
u = (x, y);
print(u);
// - but α & β can't end until here (even with NLL)
```

The assignment to `x` "ends" all borrows to things pointed to by `x` - the borrows still exists, but the alias to them through `x` is inaccessible, which means there's no need to track accesses through it.